src/Parser.elm

```diff
@@ -0,0 +1,374 @@
+import Data.AudienceFolder as F
+import Dict
+import Json.Decode as Jd
+import Set
+
+
+parse : String -> String -> Result String Data
+parse rawFolders rawAudiences =
+    case ( parseFolders rawFolders, parseAudiences rawAudiences ) of
+        ( Err err, _ ) ->
+            Err <|
+                String.concat
+                    [ "error parsing folders JSON: "
+                    , Jd.errorToString err
+                    ]
+
+        ( _, Err err ) ->
+            Err <|
+                String.concat
+                    [ "error parsing audiences JSON: "
+                    , Jd.errorToString err
+                    ]
+
+        ( Ok folders, Ok audiences ) ->
+            let
+                clean =
+                    cleanUpData folders audiences
+            in
+            case checkData clean of
+                Nothing ->
+                    Ok clean
+
+                Just err ->
+                    Err err
+
+
+checkData : Data -> Maybe String
+checkData data =
+    maybeChain
+        [ allParentsAreFolders data
+        , allParentsExist data.all (Dict.toList data.parents)
+        ]
+
+
+allParentsExist : Dict.Dict Int String -> List ( Int, Int ) -> Maybe String
+allParentsExist all parents =
+    case parents of
+        [] ->
+            Nothing
+
+        ( child, parent ) :: arents ->
+            case Dict.get parent all of
+                Nothing ->
+                    Just <|
+                        String.concat
+                            [ "parent with ID "
+                            , String.fromInt parent
+                            , " and child "
+                            , String.fromInt child
+                            , " does not exist"
+                            ]
+
+                Just _ ->
+                    allParentsExist all arents
+
+
+allParentsAreFolders : Data -> Maybe String
+allParentsAreFolders { parents, audiences } =
+    allParentsAreFoldersHelp audiences (Dict.toList parents)
+
+
+allParentsAreFoldersHelp :
+    Dict.Dict Int A.AudienceType
+    -> List ( Int, Int )
+    -> Maybe String
+allParentsAreFoldersHelp audiences parents =
+    case parents of
+        [] ->
+            Nothing
+
+        ( child, parent ) :: arents ->
+            case Dict.get parent audiences of
+                Nothing ->
+                    allParentsAreFoldersHelp audiences arents
+
+                Just _ ->
+                    Just <|
+                        String.concat
+                            [ "parent with ID "
+                            , String.fromInt parent
+                            , " and child "
+                            , String.fromInt child
+                            , " is not a folder"
+                            ]
+
+
+maybeChain : List (Maybe a) -> Maybe a
+maybeChain maybes =
+    case maybes of
+        [] ->
+            Nothing
+
+        Nothing :: aybes ->
+            maybeChain aybes
+
+        (Just m) :: _ ->
+            Just m
```

Comment:
For fun, a possible reimplementation:
```elm
maybeChain maybes =
  maybes
    |> List.filterMap identity
    |> Maybe.andThen List.head
    |> Maybe.map Just
    |> Maybe.withDefault Nothing
```

```elm
maybeChain maybes =
    List.foldl Maybe.Extra.orElse Nothing maybes
```

Reply:
These are fun. This |> style is growing on me. I think I need to use it more.